core/src/main/java/io/grpc/internal/StatsTraceContext.java

```diff
@@ -72,19 +79,56 @@ public static StatsTraceContext newServerContext(
       List<? extends ServerStreamTracer.Factory> factories,
       String fullMethodName,
       Metadata headers) {
-    if (factories.isEmpty()) {
-      return NOOP;
-    }
+    return newServerContext(factories, fullMethodName, headers, null);
+  }
+
+  /** Factory method for the server-side that accepts a {@link ServerCallStartedListener}. */
+  public static StatsTraceContext newServerContext(
+      List<? extends ServerStreamTracer.Factory> factories,
+      String fullMethodName,
+      Metadata headers,
+      ServerCallStartedListener listener) {
     StreamTracer[] tracers = new StreamTracer[factories.size()];
     for (int i = 0; i < tracers.length; i++) {
       tracers[i] = factories.get(i).newServerStreamTracer(fullMethodName, headers);
     }
-    return new StatsTraceContext(tracers);
+    return new StatsTraceContext(tracers, listener);
+  }
+
+  /**
+   * Add stream tracers from server interceptors. These are not available until after the call has
+   * begun, so if a call terminates early these tracers may never be invoked. Expected usage is
+   * that only {@link #serverFilterContext}, and potentially {@link #streamClosed}, will be called
+   * prior to this method, but it is up to the caller to enforce this ordering.
+   *
+   * <p>The newly added tracers will have their {@link ServerStreamTracer#filterContext} methods
+   * applied to the returned {@code Context}.
+   */
+  public Context.CancellableContext setServerInterceptorTracersAndFilterContext(
+      List<? extends ServerStreamTracer.Factory> factories,
+      String fullMethodName,
+      Metadata headers,
+      Context.CancellableContext context) {
+    if (!factories.isEmpty()) {
+      StreamTracer[] tracerArr = new StreamTracer[factories.size()];
+      for (int i = 0; i < tracerArr.length; i++) {
+        ServerStreamTracer tracer = factories.get(i).newServerStreamTracer(fullMethodName, headers);
+        context = tracer.filterContext(context).withCancellation();
```

Comment:
It doesn't need to be `CancellableContext` here. As far as the `ServerStreamTracer#filterContext` it _should_ just be a `Context`. This is just a result of bad 'type matching' on my part: by the time this method is invoked from `ServerImpl`, we already have a `CancellableContext` - I think it's necessary very early in the call handling process for e.g. RPC deadlines - and I was just blindly maintaining that type here.

But I'm not sure I fully understand the problem that you are describing: looking at what `withCancellation` actually does, it's obvious that it's heavier weight than we actually need but I understood your comment to mean this is actually incorrect due to a mishandling of the context's lifetime. However, since we begin with a `CancellableContext` from `ServerImpl` (created and managed by the existing code prior to this PR, so I don't think I've broken anything there, probably :)) and according to the `CancellableContext` javadoc it is "A context which inherits cancellation from its parent but which can also be independently cancelled and which will propagate cancellation to its descendants" shouldn't this actually work as-is? Since the parent `CancellableContext` from `ServerImpl.ServerTransportListenerImpl#createContext` will eventually be cancelled and that cancellation propagated to the descendants created here?

I'm worried that I'm misunderstanding something fundamental here, but the basic constraint I was operating under was that, when this method is invoked from `ServerImpl.ServerTransportListenerImpl#streamCreatedInternal`, we have a `CancellableContext` and want to get a `CancellableContext` back. The current code is just a ham-handed way of maintaining that type, but is there a reason that we don't want a `CancellableContext` at this point at all and the "conversion" from `Context` to `CancellableContext` currently done in `ServerImpl.ServerTransportListenerImpl#createContext` should be deferred until after these interceptor-provided tracers have been invoked?

Reply:
I was talking about "To avoid leaking memory, every CancellableContext must have a defined lifetime, after which it is guaranteed to be cancelled." Yes, if the parent is cancellable you are guaranteed this one will have a lifetime, although you aren't guaranteed the lifetime is appropriate (the lifetime could be the entire process lifetime in some Context cases). I think the only reason you know the lifetime is appropriate here is because you say "statsTraceCtx.serverFilterContext()" is expected to be called first, at which point the context should already be scoped to the call.

But there's costs to cancellable contexts, and it is really weird that they don't manage their own lifetime, and it's weird they are created in a loop (instead of only once at the end). I think this code should just be dealing with normal contexts; normal context input and normal context output.